provider/openstack/firewaller.go

```diff
@@ -70,16 +71,342 @@ type firewallerFactory struct {
 
 // GetFirewaller implements FirewallerFactory
 func (f *firewallerFactory) GetFirewaller(env environs.Environ) Firewaller {
-	return &defaultFirewaller{environ: env.(*Environ)}
+	return &switchingFirewaller{env: env.(*Environ)}
+}
+
+type switchingFirewaller struct {
```

Comment:
Both your idea for future changes, and leaving it to the future, sound good to me.